gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobPropCreator.java

```diff
@@ -176,15 +179,38 @@ private boolean latePathsFound(Dataset dataset) throws IOException, FileNotFound
     jobProps.setProp(MRCompactor.COMPACTION_OUTPUT_DEDUPLICATED, this.outputDeduplicated);
     jobProps.setProp(MRCompactor.COMPACTION_SHOULD_DEDUPLICATE, !this.inputDeduplicated && this.outputDeduplicated);
 
-    if (this.recompactFromOutputPaths || !MRCompactor.datasetAlreadyCompacted(this.fs, dataset)) {
-      addInputLateFilesForFirstTimeCompaction(jobProps, dataset);
+    if (this.recompactFromOutputPaths || !MRCompactor.datasetAlreadyCompacted(this.fs, dataset, renameSourceDir)) {
+      if (renameSourceDir) {
+        Set<Path> newUnrenamedDirs = MRCompactor.getInputDeepestLevelUnrenamedDirsWithFileExistence(this.fs, dataset.inputPaths());
+        if (getAllFilePathsRecursively(newUnrenamedDirs).isEmpty()) {
+          return Optional.absent();
+        }
+        LOG.info ("[{}] has unprocessed directories for first time compaction: {}", dataset.getDatasetName(), newUnrenamedDirs);
+        dataset.overwriteInputPaths(newUnrenamedDirs);
+        dataset.setRenamePaths(newUnrenamedDirs);
+      } else {
+        addInputLateFilesForFirstTimeCompaction(jobProps, dataset);
+      }
     } else {
-      Set<Path> newDataFiles = getNewDataInFolder(dataset.inputPaths(), dataset.outputPath());
-      newDataFiles.addAll(getNewDataInFolder(dataset.inputLatePaths(), dataset.outputPath()));
-      if (newDataFiles.isEmpty()) {
-        return Optional.<Dataset> absent();
+      if (renameSourceDir) {
+        Set<Path> newUnrenamedDirs = MRCompactor.getInputDeepestLevelUnrenamedDirsWithFileExistence(this.fs, dataset.inputPaths());
+        if (newUnrenamedDirs.isEmpty()) {
```

Comment:
I consolidated these two conditions and put them in one method called obtainDatasetWithJobProps